src/jit/sideeffects.h

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _SIDEEFFECTS_H_
+#define _SIDEEFFECTS_H_
+
+//------------------------------------------------------------------------
+// LclVarSet:
+//    Represents a set of lclVars. Optimized for the case that the set
+//    never holds more than a single element. This type is used internally
+//    by `AliasSet` to track the sets of lclVars that are read and
+//    written for a given alias set.
+//
+class LclVarSet final
+{
+    bool m_hasAnyLcl;
+    bool m_hasBitVector;
+
+    union
+    {
+        hashBv*  m_bitVector;
+        unsigned m_lclNum;
+    };
+
+public:
+    LclVarSet();
+
+    inline bool IsEmpty() const
+    {
+        return !m_hasAnyLcl || !m_hasBitVector || !m_bitVector->anySet();
+    }
+
+    void Add(Compiler* compiler, unsigned lclNum);
+    bool Intersects(const LclVarSet& other) const;
+    void Clear();
+};
+
+//------------------------------------------------------------------------
+// AliasSet:
+//    Represents a set of reads and writes for the purposes of alias
+//    analysis. This type partitions storage into a two categories:
+//    lclVars and addressable locations. The definition of the former is
+//    intuitive. The latter is the union of the set of address-exposed
+//    lclVars with the set of all other memory locations. Any memory
+//    access is assumed to alias any other memory access.
+//
+class AliasSet final
+{
+    bool m_readsAddressableLocation;
+    bool m_writesAddressableLocation;
+
+    LclVarSet m_lclVarReads;
+    LclVarSet m_lclVarWrites;
+
+public:
+    AliasSet();
+    AliasSet(Compiler* compiler, GenTree* node);
+
+    inline bool WritesAnyLocation() const
+    {
+        return m_writesAddressableLocation || !m_lclVarWrites.IsEmpty();
+    }
+
+    void AddNode(Compiler* compiler, GenTree* node);
+    bool InterferesWith(const AliasSet& other) const;
+    void Clear();
+};
+
+//------------------------------------------------------------------------
+// SideEffectSet:
+//    Represents a set of side effects for the purposes of analyzing code
+//    motion.
+//
+class SideEffectSet final
+{
+    unsigned m_sideEffectFlags;
+    AliasSet m_aliasSet;
```

Comment:
It would be useful to comment all the members, but in particular: comment what set of values can be in m_sideEffectFlags